vscode/quint-vscode/server/src/server.ts

```diff
@@ -26,319 +26,260 @@ import {
-    }
-    return hoverText
-  }
+      const signatureWithMarkupKind = signature?.documentation
+        ? { ...signature, documentation: { kind: 'markdown', value: signature.documentation } as MarkupContent }
+        : signature
 
-  const hoverText = [inferredDataHover(), documentationHover()]
-    .filter(s => s.length > 0)
-    .map(s => s.join('\n'))
-    .join('\n-----\n')
+      return {
+        signatures: [signatureWithMarkupKind],
+        activeSignature: 0,
+        activeParameter: 0,
+      }
+    })
 
-  if (hoverText === '') {
-    return undefined
-  }
+    connection.onDefinition((params: DefinitionParams): HandlerResult<DefinitionLink[], void> => {
+      const parsedData = this.parsedDataByDocument.get(params.textDocument.uri)
+      if (!parsedData) {
+        return []
+      }
 
-  return {
-    contents: {
-      kind: MarkupKind.Markdown,
-      value: hoverText,
-    },
-  }
-})
+      const link = findDefinition(parsedData, params.textDocument.uri, params.position)
+      if (!link) {
+        return []
+      }
 
-connection.onSignatureHelp((params: SignatureHelpParams): HandlerResult<SignatureHelp, void> => {
-  const emptySignatures = { signatures: [], activeSignature: 0, activeParameter: 0 }
+      const { nameId, name, definitionId } = link
+      if (!definitionId) {
+        return []
+      }
 
-  const parsedData = parsedDataByDocument.get(params.textDocument.uri)
-  if (!parsedData) {
-    return emptySignatures
-  }
+      const loc = parsedData.sourceMap.get(definitionId)
+      if (!loc) {
+        return []
+      }
 
-  // Find name under cursor
-  const { modules, sourceMap } = parsedData
-  const results: [Loc, bigint][] = [...sourceMap.entries()].map(([id, loc]) => [loc, id])
-  const [module, name, _scope] = findName(modules, results, params.position) ?? [undefined, undefined, undefined]
-  if (!name) {
-    return emptySignatures
-  }
+      const range = locToRange(loc)
+      const uri = URI.parse(loc.source).toString()
+
+      // Definitions start with a qualifier. This finds where the definition name
+      // actually starts and corrects the range. If the file is not opened in the
+      // editor yet, we just use the original range. In the future, we should find
+      // another way of positioning the cursor in the editor, or read the file from
+      // disk.
+      const text = documents.get(uri)?.getText(range) ?? name
+      const unqualifiedName = name.split('::').pop()!
+      const start = text.search(new RegExp(unqualifiedName))
+      return [{
+        // The range for the name being hover over
+        originSelectionRange: locToRange(parsedData.sourceMap.get(nameId)!),
+        targetUri: uri,
+        // The range for the entire definition (including the qualifier and body)
+        targetRange: range,
+        // The range for the definition's name
+        targetSelectionRange: { ...range, start: { ...range.start, character: range.start.character + start } },
+      }]
```

Comment:
We now return this `DefinitionLink` instead of a simple location, which enables the client to open or navigate to a different file if the definition is there.